File: 17_dynamic_programming/number_solitaire.py
def solution(A):
  n = len(A)

  B = [A[0]] * n

  i = 0
  for i in xrange(1, n):
    track = B[i-1]
    j = 2
    while i-j >= 0 and j < 7:
      track = max(track, B[i-j])
      j += 1
    B[i] = A[i] + track
  return B[n-1]

File: 17_dynamic_programming/number_solitaire.cpp
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> &a) {
  int n = a.size();
  vector<int> dp(n, 0);
  dp[0] = a[0];
  for (int i = 1; i < n; i++) {
    dp[i] = dp[i - 1];
    for (int j = 2; j <= min(i, 6); j++) {
      dp[i] = max(dp[i], dp[i - j]);
    }
    dp[i] += a[i];
  }
  return dp[n - 1];
}
